server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java

```diff
@@ -106,7 +106,12 @@ public final Query toQuery(QueryShardContext context) throws IOException {
                 }
             }
             if (queryName != null) {
-                context.addNamedQuery(queryName, query);
+                context.setMatchNamedQueries(true);
+                if (query instanceof SpanQuery) {
+                    query = new NamedSpanQuery(queryName, (SpanQuery)query);
+                } else {
+                    query = new NamedQuery(queryName, query);
```

Comment:
Why not use Lucene's `NamedMatches#wrapQuery` ?

Reply:
`wrapQuery` doesn't fit very well with our testing framework, unfortunately, because it produces an anonymous class that we can't get the name from.  We also need `NamedSpanQuery` because of the way we build span queries in the DSL, even though it doesn't really make sense (a bit like SpanBoostQuery)

Reply:
I understand the need for `NamedSpanQuery` but I thought that we wouldn't need to retrieve the anonymous `NamedQuery` since we can retrieve the matches directly from the `Matches` object ?

Reply:
The various query builder tests assert that the generated lucene queries are of a specific type; we can add `NamedQuery.class` or `NamedSpanQuery.class` to those type lists, but we can't do anything with an anonymous inner class.  So this is only a test issue.